src/main/java/org/breedinginsight/brapps/importer/services/BrAPIQueryService.java

```diff
@@ -0,0 +1,240 @@
+import org.brapi.client.v2.model.exceptions.ApiException;
+import org.brapi.client.v2.model.queryParams.core.ProgramQueryParams;
+import org.brapi.client.v2.modules.core.LocationsApi;
+import org.brapi.client.v2.modules.core.ProgramsApi;
+import org.brapi.client.v2.modules.germplasm.GermplasmApi;
+import org.brapi.client.v2.modules.germplasm.GermplasmAttributesApi;
+import org.brapi.client.v2.modules.phenotype.ObservationUnitsApi;
+import org.brapi.v2.model.BrAPIAcceptedSearchResponse;
+import org.brapi.v2.model.BrAPIResponse;
+import org.brapi.v2.model.BrAPIResponseResult;
+import org.brapi.v2.model.core.BrAPILocation;
+import org.brapi.v2.model.core.BrAPIProgram;
+import org.brapi.v2.model.core.request.BrAPILocationSearchRequest;
+import org.brapi.v2.model.core.response.BrAPILocationListResponse;
+import org.brapi.v2.model.core.response.BrAPIProgramListResponse;
+import org.brapi.v2.model.germ.BrAPIGermplasm;
+import org.brapi.v2.model.germ.BrAPIGermplasmAttribute;
+import org.brapi.v2.model.germ.request.BrAPIGermplasmAttributeSearchRequest;
+import org.brapi.v2.model.germ.request.BrAPIGermplasmSearchRequest;
+import org.brapi.v2.model.germ.response.BrAPIGermplasmAttributeListResponse;
+import org.brapi.v2.model.germ.response.BrAPIGermplasmListResponse;
+import org.brapi.v2.model.germ.response.BrAPIGermplasmListResponseResult;
+import org.brapi.v2.model.pheno.BrAPIObservationUnit;
+import org.brapi.v2.model.pheno.request.BrAPIObservationUnitSearchRequest;
+import org.brapi.v2.model.pheno.response.BrAPIObservationUnitListResponse;
+import org.breedinginsight.brapps.importer.model.base.Germplasm;
+import org.breedinginsight.brapps.importer.model.base.ObservationUnit;
+import org.breedinginsight.services.brapi.BrAPIClientType;
+import org.breedinginsight.services.brapi.BrAPIProvider;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.*;
+import java.util.stream.Collectors;
+
+@Singleton
+public class BrAPIQueryService {
+
+    private BrAPIProvider brAPIProvider;
+    private static Integer SEARCH_WAIT_TIME = 1000;
+    private static Integer SEARCH_TIMEOUT = Long.valueOf(TimeUnit.MINUTES.toMillis(10)).intValue();
+    private static Integer RESULTS_PER_QUERY = 10000;
+    public static String OU_ID_REFERENCE_SOURCE = "ou_id";
+
+    @Property(name = "brapi.server.reference-source")
+    private String referenceSource;
+
+    @Inject
+    public BrAPIQueryService(BrAPIProvider brAPIProvider) {
+        this.brAPIProvider = brAPIProvider;
+    }
+
+    public <T, U, V> List<V> search(Function<U, ApiResponse<Pair<Optional<T>, Optional<BrAPIAcceptedSearchResponse>>>> searchMethod,
+                               Function3<String, Integer, Integer, ApiResponse<Pair<Optional<T>, Optional<BrAPIAcceptedSearchResponse>>>> searchGetMethod,
+                               U searchBody
+    ) throws ApiException {
```

Comment:
I think that `<T,U,V>` can be reduced to `<T>`. See the `post` method below.

Reply:
Couldn't narrow it down, but you can get away with not passing the types in when you use it according to Intellij.